python/ray/autoscaler/autoscaler.py

```diff
@@ -462,18 +465,28 @@ def launch_new_node(self, count):
                 TAG_RAY_NODE_STATUS: "uninitialized",
                 TAG_RAY_LAUNCH_CONFIG: self.launch_hash,
             }, count)
+        with self.launch_lock:
+            self.num_launches_pending -= count
         if len(self.workers()) <= num_before:
             print("Warning: Num nodes failed to increase after node creation")
 
+    def launch_new_node(self, count):
+        print("StandardAutoscaler: Launching {} new nodes".format(count))
+        with self.launch_lock:
+            self.num_launches_pending += count
+        t = threading.Thread(target=self._launch_new_node, args=(count, ))
```

Comment:
So does this require the `GCPNodeProvider` and `AWSNodeProvider` class to be thread safe? Are they?

Usage of `self.provider` is not locked by `self.launch_lock`, right?

Reply:
Good point, I had just assumed they were thread safe but should have checked. Unfortunately locking ```self.provider``` would defeat the purpose of this PR. Methods of ```self.provider``` are called several times by ```StandardAutoscaler._update()``` , even if no nodes are launched/terminated, so we would effectively still be blocking on instance creation.

Note the only concurrent calls that can take place are one or more calls to create_node and at most one to another function, so the node providers do not need to be fully thread safe. In particular, both ```AWSNodeProvider``` and ```GCPNodeProvider``` are doing some internal caching right now which would be a little messy to make thread safe, but this is _not_ touched by ```create_node``` so does not need to be changed.

I think the main issue for both ```AWSNodeProvider``` and ```GCPNodeProvider``` is that ```boto3``` and ```googleclientapi``` are built on top of ```httplib``` which is not thread safe. The solution in both cases is to create thread-local sessions. Can @ericl and @hartikainen who are more familiar with these node providers than me comment on if there are any other problems with threading?

I'm a little unhappy with what the resulting API would look like. It feels weird to demand that ```create_node``` be thread safe but that other methods can be assumed to only run from the main thread. Perhaps we should at least add some check in the other methods to catch future abuses of this (similar to ```check_main_thread()``` in ```worker.py```).

Reply:
To be safe, I think we can just create a new node provider per thread via `get_node_provider`. I know this is safe for the AWS node provider at least, one thing to watch out for is that we don't leak node providers (maybe we can allocate just 1 extra node provider dedicated for launching nodes and lock around that?)

Reply:
Creating a separate node provider hadn't occurred to me, good suggestion. Took a quick look and don't think this should cause issues with GCP, maybe @hartikainen can confirm.

I've implemented this in the latest commit. There's a fixed pool of threads so this shouldn't leak node providers. (I opted to create multiple threads since there are use cases where it can be beneficial to have multiple instance requests in-flight. In particular, this will help with issue #2177.)